pci/src/device.rs

```diff
@@ -3,14 +3,20 @@
 // found in the LICENSE-BSD-3-Clause file.
 
 use crate::configuration::{self, PciConfiguration};
+use crate::msix::MsixTableEntry;
 use crate::PciInterruptPin;
 use devices::BusDevice;
 use std;
 use std::fmt::{self, Display};
+use std::sync::Arc;
 use vm_allocator::SystemAllocator;
 use vm_memory::{GuestAddress, GuestUsize};
 use vmm_sys_util::EventFd;
 
+pub type IrqClosure = Box<Fn() -> std::result::Result<(), std::io::Error> + Send + Sync>;
+pub type MsixClosure =
+    Box<Fn(MsixTableEntry) -> std::result::Result<(), std::io::Error> + Send + Sync>;
+
```

Comment:
We're up to 3 different closure types. For clarity sake, I would suggest defining:

```Rust
pub type PciInterruptClosure =
    Box<Fn(Option<MsixTableEntry>) -> std::result::Result<(), std::io::Error> + Send + Sync>;
```

and use the same type for both `assign_irq` and assign_msix`.

Reply:
Also, we need to document the closure prototype.

Reply:
Regarding your proposal about `PciInterruptClosure`, I'm not sure this is the right approach because those closures should be passed through dedicated functions (`assign_irq` or `assign_msix`) depending on the use case, hence the need to have different closures. In the context of IRQ, why would we have an `Option<MsixTableEntry>` as parameter?

The VMM (`vm.rs`) can setup different closures for its PCI devices, based on the type of interrupt they want to set, this does not need to be generic. On the other hand, `VirtioInterruptClosure` is specific to virtio and has to be generic.

Actually the thing is that `IrqClosure` could be defined somewhere else than the `pci` crate but I'm not sure where.

Reply:
Right, the IrqClosure does not have to be PCI specific and can live in another place. Probably in vm-device once it's ready. But I'm really not a fan of including type as part of type_name, e.g. from type I know this is a closure already.

Reply:
> Regarding your proposal about `PciInterruptClosure`, I'm not sure this is the right approach because those closures should be passed through dedicated functions (`assign_irq` or `assign_msix`) depending on the use case,

As a matter of fact, I even considered commenting on the `assign_msix` addition :)
What we're trying to do is extend the idea of assigning an interrupt to a device. Those interrupts can be delivered from the device in several ways: Pin based interrupts or memory based ones.

So I would go further and say we should define a generic `InterruptParameters` structure containing an `Option<MsixTableEntry>` and an `InterruptType` fields.
Then have your closure essentially being defined as:
```Rust
pub type InterruptDelivery =
    Box<Fn(InterruptParameters) -> std::result::Result<(), std::io::Error> + Send + Sync>;
```

And you and Chao are right: This shouldn't be PCI prefixed as it could potentially belong to the device model crate. For now we can keep it here.

>  hence the need to have different closures. In the context of IRQ, why would we have an `Option<MsixTableEntry>` as parameter?
> 
> The VMM (`vm.rs`) can setup different closures for its PCI devices, based on the type of interrupt they want to set, this does not need to be generic.

Yes but otoh the VMM is only trying to assign an interrupt to a device. And it wants to tell the device manager or for now the virtio layer how the device should deliver the interrupt. 

> On the other hand, `VirtioInterruptClosure` is specific to virtio and has to be generic.
> 
> Actually the thing is that `IrqClosure` could be defined somewhere else than the `pci` crate but I'm not sure where.

Reply:
Closely related discussion here: https://github.com/rust-vmm/vm-virtio/pull/4

An `DeviceInterrupt` trait seems to be the trend of the discussion, and is something we may want to consider as well.